src/cc/BPF.cc

```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright (c) 2016 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <linux/bpf.h>
+#include <unistd.h>
+#include <cstdio>
+#include <exception>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+#include "bcc_syms.h"
+#include "bpf_module.h"
+#include "common.h"
+#include "exception.h"
+#include "libbpf.h"
+#include "perf_reader.h"
+
+#include "BPF.h"
+
+namespace ebpf {
+
+std::string uint_to_hex(uint64_t value) {
+  std::stringstream ss;
+  ss << std::hex << value;
+  return ss.str();
+}
+
+std::string sanitize_str(std::string str, bool (*validator)(char),
+                         char replacement = '_') {
+  for (size_t i = 0; i < str.length(); i++)
+    if (!validator(str[i]))
+      str[i] = replacement;
+  return str;
+}
+
+BPF::BPF(const std::string& bpf_program, unsigned int debug_flag,
+         std::vector<std::string> cflags)
+    : bpf_module_(new BPFModule(debug_flag)),
+      bpf_program_(bpf_program),
+      cflags_(cflags) {}
+
+StatusTuple BPF::init() {
```

Comment:
could you move 'bpf_program' string into init() method instead of constructor ?
The constructor can act on flags that can be applied immediately,
whereas storing 'bpf_program' string just to be used in init() method is suspicious.
The user of the api will be left to wonder 'what's going on with that string during constructor'...